File: src-tauri/src/crypto/stronghold.rs
use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use iota_stronghold::{Client, KeyProvider, SnapshotPath, Stronghold as IotaStronghold};
use keyring::{Entry, Error as KeyringError};
use rand::{rngs::SysRng, TryRng};
use serde::Serialize;
use std::path::PathBuf;
use zeroize::Zeroizing;

const SERVICE_NAME: &str = "deployer-app";
const VAULT_PASSWORD_ACCOUNT: &str = "stronghold-vault";
const VAULT_FILE_NAME: &str = "vault.hold";
const CLIENT_NAME: &str = "encrypt-keys";

/// Scopes de cifrado esperados en el vault.
const EXPECTED_SCOPES: &[&str] = &[
    "encrypt:deployer_hosts.password",
    "encrypt:deployer_passkeys.key_content",
    "encrypt:deployer_passkeys.passphrase",
];

/// Reporte de salud del vault de Stronghold.
#[derive(Serialize)]
pub struct VaultHealthReport {
    pub vault_file_exists: bool,
    pub salt_file_exists: bool,
    pub snapshot_loaded: bool,
    pub client_exists: bool,
    pub keys_found: Vec<String>,
    pub expected_scopes: Vec<String>,
    pub missing_scopes: Vec<String>,
}

/// Acceso de SOLO LECTURA al vault de Stronghold para claves de cifrado.
///
/// La escritura sobre el vault (creación/rotación/purga de claves) la realiza
/// el frontend con el plugin de Stronghold. Rust solo abre el vault para leer
/// claves y descifrar credenciales SSH, de modo que no hay contención entre
/// ambos accesos.
///
/// Las claves se guardan versionadas: `encrypt:{table}.{col}:{version}`.
/// La versión actual se guarda en `encrypt:{table}.{col}:current`.
/// Un valor cifrado en SQLite tiene formato `ENC:{version}:<base64>`.
pub struct StrongholdVault {
    client: Client,
}

impl StrongholdVault {
    /// Abre el vault (solo lectura) calculando rutas desde `app_local_data_dir`.
    /// El vault se comparte con el plugin Tauri JS, que usa el mismo salt
    /// y la misma password para acceder al mismo archivo.
    ///
    /// Devuelve error si el archivo de vault no existe o si el client
    /// `encrypt-keys` no se encuentra en el snapshot (vault vacío o corrupto).
    pub fn open(app: &tauri::AppHandle) -> Result<Self, String> {
        use tauri::Manager;
        let data_dir = app
            .path()
            .app_local_data_dir()
            .map_err(|e| format!("Error al obtener directorio de datos: {}", e))?;
        let vault_path = data_dir.join(VAULT_FILE_NAME);
        let salt_path = data_dir.join("salt.txt");

        if !vault_path.exists() {
            return Err(format!(
                "Vault no encontrado en '{}'. El frontend debe inicializar el vault antes de usar credenciales cifradas. Ejecuta la app completa al menos una vez.",
                vault_path.display()
            ));
        }

        let raw_password = get_or_create_vault_password()?;
        let password_hash = hash_password(&raw_password, &salt_path)?;

        let snapshot_path = SnapshotPath::from_path(&vault_path);
        let stronghold = IotaStronghold::default();
        let keyprovider = KeyProvider::try_from(Zeroizing::new(password_hash))
            .map_err(|e| format!("Error al crear KeyProvider: {:?}", e))?;

        stronghold
            .load_snapshot(&keyprovider, &snapshot_path)
            .map_err(|e| format!("Error al cargar snapshot de Stronghold: {:?}", e))?;

        // Cargar el client 'encrypt-keys' desde el snapshot a memoria.
        // load_snapshot NO puebla la tabla de clients en memoria: hay que llamar
        // load_client para traerlo desde el snapshot, igual que hace el plugin JS
        // con loadClient antes de usarlo.
        let client = stronghold
            .load_client(CLIENT_NAME)
            .map_err(|e| {
                format!(
                    "Client 'encrypt-keys' no encontrado en el snapshot: {:?}. \
                     El vault parece estar vacío o corrupto. \
                     Verifica que el frontend ha guardado claves en el vault.",
                    e
                )
            })?;

        Ok(Self { client })
    }

    /// Realiza un diagnóstico completo del vault sin intentar descifrar.
    /// Devuelve un reporte con el estado de cada componente.
    pub fn health_check(app: &tauri::AppHandle) -> VaultHealthReport {
        use tauri::Manager;

        let mut report = VaultHealthReport {
            vault_file_exists: false,
            salt_file_exists: false,
            snapshot_loaded: false,
            client_exists: false,
            keys_found: Vec::new(),
            expected_scopes: EXPECTED_SCOPES.iter().map(|s| s.to_string()).collect(),
            missing_scopes: Vec::new(),
        };

        let data_dir = match app.path().app_local_data_dir() {
            Ok(d) => d,
            Err(_) => return report,
        };

        let vault_path = data_dir.join(VAULT_FILE_NAME);
        let salt_path = data_dir.join("salt.txt");

        report.vault_file_exists = vault_path.exists();
        report.salt_file_exists = salt_path.exists();

        if !report.vault_file_exists {
            report.missing_scopes = report.expected_scopes.clone();
            return report;
        }

        let raw_password = match get_or_create_vault_password() {
            Ok(pw) => pw,
            Err(_) => return report,
        };

        let password_hash = match hash_password(&raw_password, &salt_path) {
            Ok(h) => h,
            Err(_) => return report,
        };

        let snapshot_path = SnapshotPath::from_path(&vault_path);
        let stronghold = IotaStronghold::default();
        let keyprovider = match KeyProvider::try_from(Zeroizing::new(password_hash)) {
            Ok(kp) => kp,
            Err(_) => return report,
        };

        if stronghold.load_snapshot(&keyprovider, &snapshot_path).is_err() {
            report.missing_scopes = report.expected_scopes.clone();
            return report;
        }

        report.snapshot_loaded = true;

        let client = match stronghold.load_client(CLIENT_NAME) {
            Ok(c) => c,
            Err(_) => {
                report.missing_scopes = report.expected_scopes.clone();
                return report;
            }
        };

        report.client_exists = true;

        for &scope in EXPECTED_SCOPES {
            let mut found = false;
            for version in 0..10 {
                let key_name = format!("{}:{}", scope, version);
                if client.store().get(key_name.as_bytes()).ok().flatten().is_some() {
                    report.keys_found.push(key_name);
                    found = true;
                }
            }
            if !found {
                report.missing_scopes.push(scope.to_string());
            }
        }

        report
    }

    /// Obtiene la clave AES de una versión concreta para un scope.
    pub fn get_key_for_version(&self, scope: &str, version: i64) -> Result<Vec<u8>, String> {
        let key_scope = format!("{}:{}", scope, version);
        let key = self
            .client
            .store()
            .get(key_scope.as_bytes())
            .map_err(|e| format!("Error al leer clave del store: {:?}", e))?
            .ok_or_else(|| {
                format!(
                    "Clave no encontrada: {}. La clave no existe en el vault. \
                     Asegúrate de que el frontend ha cifrado este campo al menos una vez.",
                    key_scope
                )
            })?;
        Ok(key.to_vec())
    }

    /// Descifra un valor usando la clave de la versión con la que fue cifrado.
    ///
    /// El valor debe tener formato `ENC:{version}:<base64>`; los valores
    /// `ENC:<base64>` sin versión se tratan como versión 0.
    pub fn decrypt_value(&self, scope: &str, ciphertext: &str) -> Result<String, String> {
        let (version, payload) = split_ciphertext_version(ciphertext);
        let key = self.get_key_for_version(scope, version)?;
        decrypt_aes_gcm_payload(payload, &key)
    }
}

/// Separa la versión de un valor cifrado.
///
/// Formato: `ENC:{version}:<base64>`. Si no hay segmento de versión
/// (`ENC:<base64>`), se asume versión 0.
pub fn split_ciphertext_version(ciphertext: &str) -> (i64, &str) {
    if !ciphertext.starts_with("ENC:") {
        return (0, ciphertext);
    }

    let rest = &ciphertext[4..];

    // Buscar el primer ':' tras el prefijo. El base64 no contiene ':'.
    match rest.find(':') {
        Some(idx) if idx > 0 => {
            let version_str = &rest[..idx];
            let payload = &rest[idx + 1..];
            match version_str.parse::<i64>() {
                Ok(v) => (v, payload),
                Err(_) => (0, rest),
            }
        }
        _ => (0, rest),
    }
}

/// Descifra un payload AES-256-GCM (sin prefijo `ENC:`).
///
/// Formato del payload: `base64(nonce_12_bytes + ciphertext)`.
fn decrypt_aes_gcm_payload(payload: &str, key: &[u8]) -> Result<String, String> {
    let combined = BASE64
        .decode(payload)
        .map_err(|e| format!("Error al decodificar base64: {}", e))?;

    if combined.len() < 12 {
        return Err("Datos cifrados corruptos: longitud insuficiente".to_string());
    }

    let (nonce_bytes, encrypted_data) = combined.split_at(12);
    let nonce = Nonce::from(<[u8; 12]>::try_from(nonce_bytes).unwrap());

    let cipher = Aes256Gcm::new_from_slice(key)
        .map_err(|e| format!("Error al inicializar el cifrado: {}", e))?;

    let plaintext_bytes = cipher
        .decrypt(&nonce, encrypted_data)
        .map_err(|_| "Error al descifrar: clave incorrecta o datos corruptos".to_string())?;

    String::from_utf8(plaintext_bytes)
        .map_err(|e| format!("Error al decodificar texto descifrado: {}", e))
}

/// Obtiene la password del vault desde el keychain del SO.
/// En la primera ejecución, genera una password aleatoria y la persiste.
pub fn get_or_create_vault_password() -> Result<String, String> {
    let entry = Entry::new(SERVICE_NAME, VAULT_PASSWORD_ACCOUNT)
        .map_err(|e| format!("Error al acceder al keychain: {}", e))?;

    match entry.get_password() {
        Ok(pw) => Ok(pw),
        Err(KeyringError::NoEntry) => {
            let pw = generate_random_password();
            entry
                .set_password(&pw)
                .map_err(|e| format!("Error al guardar password del vault: {}", e))?;
            Ok(pw)
        }
        Err(e) => Err(format!("Error al leer password del vault: {}", e)),
    }
}

/// Deriva la clave de cifrado del vault con Argon2 + salt.
///
/// Mismo algoritmo que `tauri_plugin_stronghold::kdf::KeyDerivation::argon2`,
/// usando el mismo archivo de salt que el plugin.
fn hash_password(password: &str, salt_path: &PathBuf) -> Result<Vec<u8>, String> {
    let mut salt = [0u8; 32];

    if salt_path.exists() {
        let data = std::fs::read(salt_path)
            .map_err(|e| format!("Error al leer salt: {}", e))?;
        if data.len() != 32 {
            return Err("Archivo de salt corrupto: longitud incorrecta".to_string());
        }
        salt.copy_from_slice(&data);
    } else {
        if let Some(parent) = salt_path.parent() {
            std::fs::create_dir_all(parent)
                .map_err(|e| format!("Error al crear directorio del salt: {}", e))?;
        }
        SysRng
            .try_fill_bytes(&mut salt)
            .map_err(|e| format!("Error al generar salt: {}", e))?;
        std::fs::write(salt_path, &salt)
            .map_err(|e| format!("Error al guardar salt: {}", e))?;
    }

    argon2::hash_raw(password.as_bytes(), &salt, &Default::default())
        .map_err(|e| format!("Error al derivar clave con Argon2: {:?}", e))
}

fn generate_random_password() -> String {
    let mut bytes = [0u8; 32];
    SysRng.try_fill_bytes(&mut bytes).ok();
    hex::encode(bytes)
}

File: src/drizzle/lib/stronghold.ts
import { Stronghold } from '@tauri-apps/plugin-stronghold'
import { invoke } from '@tauri-apps/api/core'

// ====================================================================
// Cripto con Stronghold (claves versionadas, estilo Symfony)
//
// El vault se gestiona íntegramente desde el frontend con el plugin
// @tauri-apps/plugin-stronghold (única autoridad de escritura). Las
// claves AES se guardan como:
//   encrypt:{table}.{col}:{version}  → 32 bytes (clave AES-256-GCM)
//   encrypt:{table}.{col}:current    → 8 bytes big-endian (versión actual)
//
// El cifrado/descifrado del contenido se hace con Web Crypto (AES-GCM),
// porque Stronghold no provee AES sobre datos externos.
//
// Rust SOLO lee el vault (sin escribir) para descifrar credenciales SSH,
// de modo que no hay contención entre ambos accesos.
//
// Valor en SQLite: ENC:{version}:<base64(nonce12+cipher)>
// (un valor ENC:<base64> sin versión se trata como versión 0)
// ====================================================================

const ENC_PREFIX = 'ENC:'
const CURRENT_VERSION_SUFFIX = ':current'
const CLIENT_NAME = 'encrypt-keys'

let strongholdPromise: Promise<Stronghold> | null = null
let clientPromise: Promise<import('@tauri-apps/plugin-stronghold').Client> | null = null
const keyCache = new Map<string, CryptoKey>()

/**
 * Genera el scope de cifrado con la tabla incluida: `encrypt:{table}.{field}`.
 * Incluir la tabla evita compartir clave AES entre tablas que tengan columnas
 * con el mismo nombre.
 */
export function encryptScope(table: string, field: string): string {
	return `encrypt:${table}.${field}`
}

/**
 * Estado del vault de Stronghold reportado por el backend (diagnóstico).
 */
export interface VaultHealthReport {
	vault_file_exists: boolean
	salt_file_exists: boolean
	snapshot_loaded: boolean
	client_exists: boolean
	keys_found: string[]
	expected_scopes: string[]
	missing_scopes: string[]
}

/**
 * Solicita al backend un diagnóstico del vault de Stronghold.
 * Permite detectar si el vault está vacío, corrupto o sin inicializar.
 */
export async function checkVaultHealth(): Promise<VaultHealthReport> {
	return invoke<VaultHealthReport>('check_vault_health')
}

async function getStronghold(): Promise<Stronghold> {
	if (!strongholdPromise) {
		strongholdPromise = (async () => {
			const vaultPath = await invoke<string>('get_vault_path')
			const password = await invoke<string>('get_vault_password')
			return await Stronghold.load(vaultPath, password)
		})()
	}
	return strongholdPromise
}

async function getStore(): Promise<import('@tauri-apps/plugin-stronghold').Store> {
	if (!clientPromise) {
		clientPromise = (async () => {
			const stronghold = await getStronghold()
			try {
				return await stronghold.loadClient(CLIENT_NAME)
			} catch {
				return await stronghold.createClient(CLIENT_NAME)
			}
		})()
	}
	return (await clientPromise).getStore()
}

function storeKey(scope: string, version: number): string {
	return `${scope}:${version}`
}

function currentKey(scope: string): string {
	return `${scope}${CURRENT_VERSION_SUFFIX}`
}

function bytesToInt(bytes: Uint8Array): number {
	const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
	return Number(view.getBigInt64(0, false))
}

function intToBytes(n: number): Uint8Array {
	const arr = new Uint8Array(8)
	new DataView(arr.buffer).setBigInt64(0, BigInt(n), false)
	return arr
}

/**
 * Obtiene (o crea en v0, migrando del formato legado) la versión actual
 * de un scope. Guarda las claves en el store de Stronghold.
 */
async function ensureCurrentVersion(scope: string): Promise<number> {
	const store = await getStore()

	const raw = await store.get(currentKey(scope))
	if (raw && raw.byteLength === 8) {
		return bytesToInt(raw)
	}

	// No hay versión actual → migrar de formato legado o crear v0.
	const version = 0
	const legacy = await store.get(scope)

	if (legacy && legacy.byteLength === 32) {
		await store.insert(storeKey(scope, version), Array.from(legacy))
		await store.remove(scope)
	} else {
		await store.insert(storeKey(scope, version), Array.from(crypto.getRandomValues(new Uint8Array(32))))
	}

	await store.insert(currentKey(scope), Array.from(intToBytes(version)))
	const stronghold = await getStronghold()
	await stronghold.save()

	return version
}

/**
 * Obtiene la versión actual de cifrado para un scope.
 */
export async function getCurrentVersion(scope: string): Promise<number> {
	return ensureCurrentVersion(scope)
}

/**
 * Obtiene (o crea) la CryptoKey AES-256-GCM de una versión concreta.
 * Si `version` es undefined, usa la versión actual.
 */
async function getCryptoKey(scope: string, version?: number): Promise<{ key: CryptoKey; version: number }> {
	const v = version ?? (await ensureCurrentVersion(scope))

	const cacheKey = `${scope}:${v}`
	const cached = keyCache.get(cacheKey)
	if (cached) return { key: cached, version: v }

	const store = await getStore()
	const rawKey = await store.get(storeKey(scope, v))

	let keyBytes: Uint8Array
	if (rawKey && rawKey.byteLength === 32) {
		keyBytes = rawKey
	} else {
		keyBytes = crypto.getRandomValues(new Uint8Array(32))
		await store.insert(storeKey(scope, v), Array.from(keyBytes))
		const stronghold = await getStronghold()
		await stronghold.save()
	}

	const keyBuffer = keyBytes.buffer.slice(keyBytes.byteOffset, keyBytes.byteOffset + keyBytes.byteLength) as ArrayBuffer
	const cryptoKey = await crypto.subtle.importKey('raw', keyBuffer, { name: 'AES-GCM', length: 256 }, false, [
		'encrypt',
		'decrypt',
	])

	keyCache.set(cacheKey, cryptoKey)
	return { key: cryptoKey, version: v }
}

function toBase64(bytes: Uint8Array): string {
	let binary = ''
	for (let i = 0; i < bytes.length; i++) {
		binary += String.fromCharCode(bytes[i]!)
	}
	return btoa(binary)
}

function fromBase64(b64: string): Uint8Array {
	const binary = atob(b64)
	const bytes = new Uint8Array(binary.length)
	for (let i = 0; i < binary.length; i++) {
		bytes[i] = binary.charCodeAt(i)
	}
	return bytes
}

/**
 * Separa la versión de un valor cifrado.
 * Formato: `ENC:{version}:<base64>`. `ENC:<base64>` sin versión → versión 0.
 */
export function splitCiphertextVersion(ciphertext: string): { version: number; payload: string } {
	if (!ciphertext.startsWith(ENC_PREFIX)) return { version: 0, payload: ciphertext }

	const rest = ciphertext.slice(ENC_PREFIX.length)
	const sep = rest.indexOf(':')
	if (sep > 0) {
		const parsed = Number(rest.slice(0, sep))
		if (Number.isInteger(parsed) && parsed >= 0) {
			return { version: parsed, payload: rest.slice(sep + 1) }
		}
	}
	return { version: 0, payload: rest }
}

/**
 * Cifra un texto plano con AES-256-GCM usando la versión actual de la clave.
 * Resultado: `ENC:{version}:<base64(nonce_12_bytes + ciphertext)>`.
 * Si el valor ya está cifrado, lo devuelve tal cual (no re-cifra).
 */
export async function encrypt(scope: string, plaintext: string): Promise<string> {
	if (plaintext.startsWith(ENC_PREFIX)) return plaintext

	const { key, version } = await getCryptoKey(scope)
	const iv = crypto.getRandomValues(new Uint8Array(12))
	const ciphertext = await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, new TextEncoder().encode(plaintext))

	const combined = new Uint8Array(12 + ciphertext.byteLength)
	combined.set(iv, 0)
	combined.set(new Uint8Array(ciphertext), 12)

	return `${ENC_PREFIX}${version}:${toBase64(combined)}`
}

/**
 * Descifra un valor cifrado usando la clave de la versión con la que fue
 * cifrado (las versiones antiguas se conservan). Si no tiene prefijo
 * `ENC:`, se devuelve tal cual.
 */
export async function decrypt(scope: string, ciphertext: string): Promise<string> {
	if (!ciphertext.startsWith(ENC_PREFIX)) return ciphertext

	const { version, payload } = splitCiphertextVersion(ciphertext)
	const { key } = await getCryptoKey(scope, version)

	const combined = fromBase64(payload)
	if (combined.length < 12) {
		throw new Error('Datos cifrados corruptos: longitud insuficiente')
	}

	const iv = combined.slice(0, 12)
	const plaintext = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, combined.slice(12))

	return new TextDecoder().decode(plaintext)
}

/**
 * Rota la clave AES de un scope.
 * Crea la clave `current+1` y actualiza el índice `current`, conservando
 * la clave anterior (los valores cifrados con versiones viejas siguen
 * descifrándose). Devuelve la nueva versión y los bytes de la clave.
 */
export async function rotateKey(scope: string): Promise<{ version: number; keyBytes: Uint8Array }> {
	const current = await ensureCurrentVersion(scope)
	const newVersion = current + 1
	const newKey = crypto.getRandomValues(new Uint8Array(32))

	const store = await getStore()
	await store.insert(storeKey(scope, newVersion), Array.from(newKey))
	await store.insert(currentKey(scope), Array.from(intToBytes(newVersion)))
	const stronghold = await getStronghold()
	await stronghold.save()

	const keyBuffer = newKey.buffer.slice(newKey.byteOffset, newKey.byteOffset + newKey.byteLength) as ArrayBuffer
	const cryptoKey = await crypto.subtle.importKey('raw', keyBuffer, { name: 'AES-GCM', length: 256 }, false, [
		'encrypt',
		'decrypt',
	])
	keyCache.set(`${scope}:${newVersion}`, cryptoKey)

	return { version: newVersion, keyBytes: newKey }
}

/**
 * Provee la lista de campos cifrados (scope y etiqueta) para la UI.
 */
export const ENCRYPTED_FIELDS = [
	{ scope: encryptScope('deployer_hosts', 'password'), label: 'deployer_hosts.password', table: 'deployer_hosts', field: 'password' },
	{ scope: encryptScope('deployer_passkeys', 'key_content'), label: 'deployer_passkeys.key_content', table: 'deployer_passkeys', field: 'key_content' },
	{ scope: encryptScope('deployer_passkeys', 'passphrase'), label: 'deployer_passkeys.passphrase', table: 'deployer_passkeys', field: 'passphrase' },
] as const

/**
 * Purga las versiones antiguas de un scope que ya no se usan en ningún valor
 * y que aún existen en el store. Solo cuenta las que se purgan de verdad.
 */
async function purgeUnusedVersions(scope: string, currentVersion: number, used: [string, number][]): Promise<string[]> {
	const purged: string[] = []
	const store = await getStore()
	for (let version = 1; version < currentVersion; version++) {
		const stillUsed = used.some(([s, v]) => s === scope && v === version)
		if (!stillUsed) {
			const existing = await store.get(storeKey(scope, version))
			if (existing) {
				await store.remove(storeKey(scope, version))
				purged.push(`${scope} (v${version})`)
			}
		}
	}
	// solo guardar si realmente se quitaron claves
	if (purged.length > 0) {
		const stronghold = await getStronghold()
		await stronghold.save()
	}
	return purged
}

interface QueryRawResponse {
	success: boolean
	data: { columns: string[]; rows: unknown[][] } | null
}

interface ScanResult {
	reencrypted: number
	purged_versions: string[]
	versions_in_use: number
}

/**
 * Recifra con la versión actual todos los valores de los campos dados y purga
 * las versiones antiguas sin uso de sus scopes.
 */
async function reencryptFields(fields: readonly (typeof ENCRYPTED_FIELDS)[number][]): Promise<ScanResult> {
	let reencrypted = 0
	const used: [string, number][] = []
	const updates: { field: (typeof ENCRYPTED_FIELDS)[number]; pk: unknown; encrypted: string }[] = []

	for (const field of fields) {
		const currentVersion = await ensureCurrentVersion(field.scope)
		const selectSql = `SELECT "id", "${field.field}" FROM "${field.table}"`

		const response = await invoke<QueryRawResponse>('query_raw', {
			sql: selectSql,
			params: [],
		})
		if (!response.success || !response.data) continue

		const rows = response.data.rows

		for (const row of rows) {
			const pk = row[0]
			const value = typeof row[1] === 'string' ? row[1] : null
			if (!value || !value.startsWith(ENC_PREFIX)) continue

			const { version } = splitCiphertextVersion(value)
			used.push([field.scope, version])

			if (version !== currentVersion) {
				const currentKey = await getCryptoKey(field.scope, currentVersion)
				const plaintext = await decrypt(field.scope, value)

				if (plaintext.startsWith(ENC_PREFIX)) {
					// no se pudo descifrar por falta de clave antigua
					continue
				}

				const iv = crypto.getRandomValues(new Uint8Array(12))
				const ciphertext = await crypto.subtle.encrypt(
					{ name: 'AES-GCM', iv },
					currentKey.key,
					new TextEncoder().encode(plaintext),
				)
				const combined = new Uint8Array(12 + ciphertext.byteLength)
				combined.set(iv, 0)
				combined.set(new Uint8Array(ciphertext), 12)

				updates.push({
					field,
					pk,
					encrypted: `${ENC_PREFIX}${currentVersion}:${toBase64(combined)}`,
				})
			}
		}
	}

	// Aplicar re-cifrado (escritura directa — el proxy de db.ts re-cifraría, así que
	// usamos query_raw directamente con el valor ya cifrado).
	for (const u of updates) {
		const updateSql = `UPDATE "${u.field.table}" SET "${u.field.field}" = ? WHERE "id" = ?`
		const response = await invoke<QueryRawResponse>('query_raw', {
			sql: updateSql,
			params: [u.encrypted, u.pk],
		})
		if (response.success) reencrypted++
	}

	// Purgar versiones antiguas sin uso de todos los scopes procesados.
	const purged_versions: string[] = []
	for (const field of fields) {
		const currentVersion = await ensureCurrentVersion(field.scope)
		purged_versions.push(...(await purgeUnusedVersions(field.scope, currentVersion, used)))
	}

	keyCache.clear()

	return { reencrypted, purged_versions, versions_in_use: used.filter(([, v]) => v > 0).length }
}

/**
 * Escanea TODOS los valores cifrados de todos los scopes, re-cifra los que
 * usan una versión anterior a la actual y purga las claves viejas sin uso.
 */
export async function scanAndReencrypt(): Promise<ScanResult> {
	return reencryptFields(ENCRYPTED_FIELDS)
}

/**
 * Recifra todos los valores de un scope con su versión actual y purga las
 * versiones antiguas sin uso. Se usa tras rotar la clave, de modo que los
 * valores pasan de inmediato a la clave nueva.
 */
export async function reencryptScope(scope: string): Promise<ScanResult> {
	const fields = ENCRYPTED_FIELDS.filter((f) => f.scope === scope)
	return reencryptFields(fields)
}
